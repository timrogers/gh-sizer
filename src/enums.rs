use std::fmt;

#[derive(clap::ValueEnum, Clone, Debug)]
pub enum OutputFormat {
    Text,
    Json,
}

impl fmt::Display for OutputFormat {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            OutputFormat::Text => write!(f, "text"),
            OutputFormat::Json => write!(f, "json"),
        }
    }
}
