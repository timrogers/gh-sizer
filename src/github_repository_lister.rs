use std::io::Error;
use std::process::Command;

#[cfg(test)]
use mockall::{automock, predicate::*};

#[cfg_attr(test, automock)]
pub trait GitHubRepositoryLister {
    fn call(&self, owner: &str) -> Result<Vec<String>, Error>;
}

pub struct GitHubRepositoryListerImpl;

impl GitHubRepositoryLister for GitHubRepositoryListerImpl {
    fn call(&self, owner: &str) -> Result<Vec<String>, Error> {
        let mut list_command = Command::new("gh");
        list_command.arg("repo");
        list_command.arg("list");
        list_command.arg(owner);
        list_command.arg("-L");
        list_command.arg("10000");
        list_command.arg("--json");
        list_command.arg("name");
        list_command.arg("--jq");
        list_command.arg(".[].name");

        let list_command_output = list_command.output()?;

        if !list_command_output.status.success() {
            eprintln!("{}", String::from_utf8_lossy(&list_command_output.stderr));
            std::process::exit(exitcode::DATAERR);
        }

        let output_text = String::from_utf8_lossy(&list_command_output.stdout);
        let output_lines: Vec<String> = output_text
            .split_whitespace()
            .map(|s| s.to_string())
            .collect();

        Ok(output_lines)
    }
}
