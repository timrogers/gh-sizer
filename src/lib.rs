const VERSION: &str = env!("CARGO_PKG_VERSION");

pub mod enums;
pub use crate::enums::OutputFormat;

pub mod github_repository_lister;
pub use crate::github_repository_lister::GitHubRepositoryLister;

pub mod generate_script {
    use crate::enums::OutputFormat;
    use crate::enums::ScriptType;
    use crate::github_repository_lister::GitHubRepositoryLister;
    use std::io::{Error, Write};
    use std::path::PathBuf;

    #[cfg(test)]
    use crate::github_repository_lister::MockGitHubRepositoryLister;

    fn generate_powershell_script(
        repository_names: Vec<String>,
        owner: &str,
        output_format: OutputFormat,
        output_directory: &str,
        output_filename_template: &str,
        gh_sizer_command: &str,
    ) -> String {
        let mut generated_script: String = format!(
            "#!/usr/bin/env pwsh\n\n# Script generated by gh-sizer v{}\n\n",
            crate::VERSION
        );

        // Update line to create a directory
        let script_line = format!(
            "New-Item -Name {} -ItemType directory -Force | Out-Null\n",
            output_directory
        )
        .to_string();
        generated_script.push_str(&script_line);

        for repository_name in repository_names.iter() {
            let output_filename = output_filename_template
                .replace("${owner}", owner)
                .replace("${repository}", repository_name);
            let output_path = PathBuf::from(output_directory).join(output_filename);
            let script_line = format!(
                "{} repo {}/{} --output-format {} | Out-File -Path {}\n",
                gh_sizer_command,
                owner,
                repository_name,
                output_format,
                output_path.display()
            )
            .to_string();
            generated_script.push_str(&script_line);
        }

        generated_script.push_str(&format!(
            "\n# Script generated by gh-sizer v{}",
            crate::VERSION
        ));

        return generated_script;
    }

    fn generate_bash_script(
        repository_names: Vec<String>,
        owner: &str,
        output_format: OutputFormat,
        output_directory: &str,
        output_filename_template: &str,
        gh_sizer_command: &str,
    ) -> String {
        let mut generated_script: String = format!(
            "#!/usr/bin/env bash\n\n# Script generated by gh-sizer v{}\n\n",
            crate::VERSION
        );

        let script_line = format!("mkdir -p {}\n", output_directory).to_string();
        generated_script.push_str(&script_line);

        for repository_name in repository_names.iter() {
            let output_filename = output_filename_template
                .replace("${owner}", owner)
                .replace("${repository}", repository_name);
            let output_path = PathBuf::from(output_directory).join(output_filename);
            let script_line = format!(
                "{} repo {}/{} --output-format {} > {}\n",
                gh_sizer_command,
                owner,
                repository_name,
                output_format,
                output_path.display()
            )
            .to_string();
            generated_script.push_str(&script_line);
        }

        generated_script.push_str(&format!(
            "\n\n# Script generated by gh-sizer v{}",
            crate::VERSION
        ));

        return generated_script;
    }

    pub fn call(
        owner: &str,
        script_type: ScriptType,
        output_format: OutputFormat,
        output_directory: &str,
        output_filename_template: &str,
        gh_sizer_command: &str,
        github_repository_lister: &impl GitHubRepositoryLister,
        stderr: &mut impl Write,
    ) -> Result<String, Error> {
        let repository_names = github_repository_lister.call(owner)?;

        writeln!(
            stderr,
            "Generating script for {} repositories owned by {}",
            repository_names.len(),
            owner
        )?;

        let generated_script;

        if matches!(script_type, ScriptType::Bash) {
            generated_script = generate_bash_script(
                repository_names,
                owner,
                output_format,
                output_directory,
                output_filename_template,
                gh_sizer_command,
            );
        } else {
            generated_script = generate_powershell_script(
                repository_names,
                owner,
                output_format,
                output_directory,
                output_filename_template,
                gh_sizer_command,
            );
        }

        Ok(generated_script)
    }

    #[test]
    #[cfg(not(windows))]
    fn generate_script_generates_bash_script_with_file_output() {
        let mut lister_mock = MockGitHubRepositoryLister::new();

        lister_mock
            .expect_call()
            .returning(|_| Ok(vec!["gh-sizer".to_string()]));

        let mut stderr = Vec::new();

        let bash_script = call(
            "github",
            ScriptType::Bash.to_owned(),
            OutputFormat::Text.to_owned(),
            "output/directory",
            "${owner}-${repository}.txt",
            "gh sizer",
            &lister_mock,
            &mut stderr,
        )
        .unwrap();

        insta::assert_yaml_snapshot!(bash_script);
    }

    #[test]
    #[cfg(not(windows))]
    fn generate_script_generates_powershell_script_with_file_output() {
        let mut lister_mock = MockGitHubRepositoryLister::new();

        lister_mock
            .expect_call()
            .returning(|_| Ok(vec!["gh-sizer".to_string()]));

        let mut stderr = Vec::new();

        let powershell_script = call(
            "github",
            ScriptType::Powershell.to_owned(),
            OutputFormat::Text.to_owned(),
            "output/directory",
            "${owner}-${repository}.txt",
            "gh sizer",
            &lister_mock,
            &mut stderr,
        )
        .unwrap();

        insta::assert_yaml_snapshot!(powershell_script);
    }

    #[test]
    #[cfg(not(windows))]
    fn generate_script_generates_bash_script_with_json_formatted_output() {
        let mut lister_mock = MockGitHubRepositoryLister::new();

        lister_mock
            .expect_call()
            .returning(|_| Ok(vec!["gh-sizer".to_string()]));

        let mut stderr = Vec::new();

        let bash_script = call(
            "github",
            ScriptType::Bash.to_owned(),
            OutputFormat::Json.to_owned(),
            "output/directory",
            "${repository}.txt",
            "gh sizer",
            &lister_mock,
            &mut stderr,
        )
        .unwrap();

        insta::assert_yaml_snapshot!(bash_script);
    }

    #[test]
    #[cfg(not(windows))]
    fn generate_script_generates_powershell_script_with_json_formatted_output() {
        let mut lister_mock = MockGitHubRepositoryLister::new();

        lister_mock
            .expect_call()
            .returning(|_| Ok(vec!["gh-sizer".to_string()]));

        let mut stderr = Vec::new();

        let powershell_script = call(
            "github",
            ScriptType::Powershell.to_owned(),
            OutputFormat::Json.to_owned(),
            "output/directory",
            "${repository}.txt",
            "gh sizer",
            &lister_mock,
            &mut stderr,
        )
        .unwrap();

        insta::assert_yaml_snapshot!(powershell_script);
    }

    #[test]
    fn generate_script_logs_number_of_repos() {
        let mut lister_mock = MockGitHubRepositoryLister::new();

        lister_mock
            .expect_call()
            .returning(|_| Ok(vec!["gh-sizer".to_string()]));

        let mut stderr = Vec::new();

        call(
            "github",
            ScriptType::Bash.to_owned(),
            OutputFormat::Json.to_owned(),
            "output/directory",
            "${repository}.txt",
            "gh sizer",
            &lister_mock,
            &mut stderr,
        )
        .unwrap();

        assert_eq!(
            String::from_utf8_lossy(&stderr),
            "Generating script for 1 repositories owned by github\n"
        );
    }
}
