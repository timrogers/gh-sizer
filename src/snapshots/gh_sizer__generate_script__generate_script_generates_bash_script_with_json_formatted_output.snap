---
source: src/lib.rs
expression: bash_script
---
"#!/usr/bin/env bash\n\n# Script generated by gh-sizer v0.1.0\n\nmkdir -p output/directory\ngh sizer repo github/gh-sizer --output-format json > output/directory/gh-sizer.txt\n\n\n# Script generated by gh-sizer v0.1.0"

